### DISCUSSION TITLE
Override provider configuration

### DISCUSSION THREAD

**Discussion by jmgilman** [gruntwork-io/knowledge-base#621] [How do I]
Assume I have a root `terragrunt.hcl` with:

```hcl
generate "provider" {
  path = "provider.tf"
  if_exists = "overwrite"
  contents = <<EOF
provider "aws" {
  region = "eu-central-1"
}
EOF
}
```

In a child module, I want to configure a specific IAM role to use, so in `terragrunt.hcl`:

```hcl
include "root" {
  path = find_in_parent_folders()
}
```

And in `main.tf`:

```hcl
provider "aws" {
  assume_role {
    role_arn     = "arn:aws:iam::123456789012:role/ROLE_NAME"
  }
}
```

This won't work because Terraform complains about duplicate provider configurations. How would I override the root configuration to specify a role ARN in this case?

There's a [similar question on StackOverflow](https://stackoverflow.com/questions/74072351/terragrunt-is-possible-to-overwrite-a-generate-block-from-child-file) that has gone unanswered.

---

<ins datetime="2022-12-21T23:27:22Z">
  <p><a href="https://support.gruntwork.io/hc/requests/109744">Tracked in ticket #109744</a></p>
</ins>

**Comment by denis256** [comment]
Hi,
for case of `role_arn` can be used [`iam_role`](https://terragrunt.gruntwork.io/docs/reference/config-blocks-and-attributes/#iam_role) field in child HCL file.

In case if different parameters are required to be injected, custom configuration can be loaded through `read_terragrunt_config`

Full example in https://github.com/denis256/terragrunt-tests/tree/master/discussion-621

```
# parent terragrunt.hcl
locals {
  provider_conf = read_terragrunt_config("provider.hcl", {inputs = { config = ""}})
}
generate "provider" {
  path = "provider.tf"
  if_exists = "overwrite"
  contents = <<EOF
provider "aws" {
  region = "eu-central-1"
  ${local.provider_conf.inputs.config}
}
EOF
}

# app2/provider.hcl
inputs = {
  config = <<EOF
  #
  # custom provider config
  #
  iam_role = "arn:aws:iam::ACCOUNT_ID:role/ROLE_NAME"
  EOF
}

# app2/terragrunt.hcl
include "root" {
  path = find_in_parent_folders()
}

# app2/provider.tf
# Generated by Terragrunt. Sig: nIlQXj57tbuaRZEa
provider "aws" {
  region = "eu-central-1"
  #
  # custom provider config
  #
  iam_role = "arn:aws:iam::ACCOUNT_ID:role/ROLE_NAME"
}
```

Most probably exists other approaches too - I shared some ideas
